packages/databyss-services/blocks/joins.ts

```diff
@@ -50,3 +53,57 @@ export const getBlocksInPages = <T extends Block>(
     .filter((b) => Boolean(b)) as T[]
   return blocks
 }
+
+export const getBlocksFromBlockRelations = <T extends Block>(
+  blockRelationDict: DocumentDict<BlockRelationResponse>,
+  blockDict: DocumentDict<Block>,
+  pageDict: DocumentDict<Page>,
+  includeArchived: boolean
+) => {
+  const filtered = joinBlockRelations({
+    blockRelationDict,
+    pageDict,
+    pagePredicate: (page) => Boolean(page.archive) === includeArchived,
+  })
+
+  const blocks = Object.values(filtered)
+    .map((blockRelation) => blockDict[blockRelation._id.substring(2)])
+    .filter((b) => Boolean(b)) as T[]
+  return blocks
+}
+
+const populatePage = ({
+  page,
+  blocks,
+}: {
+  page: Page
+  blocks: DocumentDict<Block>
+}) => {
+  const _page = page
+  _page.blocks = page.blocks.map((b) => blocks[b._id])
+  return _page
+}
+
+export const addPagesToBlockRelation = ({
```

Comment:
This could be very slow for large pages with lots of relations, so we should leverage `react-query` to manage caching on this. The idea is that the Index pages which currently call this function before rendering would actually just render an array of `IndexedPage` components. In `IndexedPage`, we would call a `useIndexedPage(pageId, relatedBlockId)`, which would do what `indexPage()` is doing now but with a cache query key like `['indexedPage', pageId, relatedBlockId]`. That way, we can intelligently clear (and even rebuild ahead of time) the cache whenever we edit the page.